numba/targets/arraymath.py

```diff
@@ -3541,6 +3542,32 @@ def np_extract_impl(condition, arr):
 
     return np_extract_impl
 
+  
+@overload(np.select)
+def np_select(condlist, choicelist, default=0):
+
+    def np_select_impl(condlist, choicelist, default=0):
+        out = default * np.ones(choicelist[0].shape, choicelist[0].dtype)
+        #should use reversed+zip, but reversed is not available
+        for i in range(len(condlist) - 1, -1, -1):
+            cond = condlist[i]
+            choice = choicelist[i]
+            out = np.where(cond, choice, out)
+        return out
+
+    if ((isinstance(condlist, (types.List, types.UniTuple)))
+        and (isinstance(choicelist, (types.List, types.UniTuple)))
+        and isinstance(condlist[0], types.Array)
+        and isinstance(condlist[0].dtype, types.Boolean)
+        and isinstance(default, (int, types.scalars.Number))):
```

Comment:
hmmm, seems like `default` can also be a boolean so `types.Boolean` and `bool` should probably also be accepted for the type of that argument.

Reply:
on catching the problems individually, do you mean that instead of having one `if` statement that checks all types and that when it fails it just returns the automatic type mismatch error message, we should have individual `if` statements that do the type checking and return a customized error message? Something like this?
```python
    if not isinstance(condlist, (types.List, types.UniTuple)):
        raise TypeError('condlist must be a List or a Tuple')
        
    elif not isinstance(choicelist, (types.List, types.UniTuple)):
        raise TypeError('choicelist must be a List or a Tuple')

    ...more checks...
    else:
        return np_select_impl
```

Reply:
about booleans, is `bool` different from `types.Boolean`? Should I always add both to the instance check?

Reply:
> on catching the problems individually, do you mean that instead of having one `if` statement that checks all types and that when it fails it just returns the automatic type mismatch error message, we should have individual `if` statements that do the type checking and return a customized error message? Something like this?
> 
> ```python
>     if not isinstance(condlist, (types.List, types.UniTuple)):
>         raise TypeError('condlist must be a List or a Tuple')
>         
>     elif not isinstance(choicelist, (types.List, types.UniTuple)):
>         raise TypeError('choicelist must be a List or a Tuple')
> 
>     ...more checks...
>     else:
>         return np_select_impl
> ```

yes, something like that quoted would be good, because it tells the user precisely which argument is problematic.

Reply:
> about booleans, is `bool` different from `types.Boolean`? Should I always add both to the instance check?

I don't think you'd be able to get a `bool` passed in to the type inference part of the code, could you please provide some context? `bool` and `types.Boolean` are different, the former is the Python `bool` the latter is the Numba type system type for `bool`.

Reply:
regarding `bool`, you said 
> so types.Boolean and bool should probably also be accepted

so I thought you meant that 

`isinstance(condlist[0].dtype, types.Boolean)` is not enough and I should add something for `bool`. Did I misunderstand?

Reply:
Oh, apologies, I might have been thinking about a inferred boolean literal (not working yet, so don't worry!), or  I possibly just missed a comma in the original sentence which made it ambiguous. I think just checking against `types.Boolean` should be fine. Thanks.